### PR TITLE
M3-3966 PromotionalOffer component

### PR DESCRIPTION
**Pull request by johnwcallahan**
## Description

**Note:** This is still WIP, but I'm looking for feedback on the data shape, API, pattern, etc.

This PR introduces a new pattern for displaying promotional offers in regard to specific features (not account promotions). The idea is that the actual content of the offer can be controlled with feature flags, using the following interface:

``` typescript
interface PromotionalOffer {
  name: string;
  body: string;
  footnote: string;
  logo: string;
  alt: string;
  backgroundColor: string;
  bodyColor: string;
  footnoteColor: string;
  features: PromotionalOfferFeature[];
  displayOnDashboard: boolean;
}

type PromotionalOfferFeature =
  | 'Linodes'
  | 'Volumes'
  | 'NodeBalancers'
  | 'Object Storage'
  | 'Kubernetes';
```

Before launching a new promotion, Cloud needs to verify:

- The asset is included in src/assets/icons/promotionalOffers and has a mapping.
- The “feature” attribute corresponds to a feature that has the logic to display a promotion on its landing page (in this PR, this logic has been added for Object Storage).

It’s not ideal to have to do a release to upload assets. I thought about using` ` but that has security implications that would need to be dealt with. I feel the most comfortable with Cloud being in control of this.

 Eventually we might have an actual CMS to handle this, so this could be somewhat of a short-term fix (or not).

**TODO:**
- [x] ~Primary Nav indicator (see ticket)~ this got canned.
- [x] Take into account `displayOnDashboard` and `feature` for conditional rendering
- [x] Tests (waiting for stakeholder approval of the design/layout before doing this)

## Type of Change
- Non breaking change ('update', 'change')

### CODE REVIEW

File: packages/manager/src/features/Dashboard/Dashboard.tsx [hunk 1]
```diff
@@ -126,6 +139,13 @@ export const Dashboard: React.StatelessComponent<CombinedProps> = props => {
         </Grid>
         <Grid item xs={12} md={5}>
           <TransferDashboardCard />
+
+          {promotionalOffers.map(promotionalOffer => (
```
Comment (johnwcallahan) [inline packages/manager/src/features/Dashboard/Dashboard.tsx]: TODO: This needs to take in to account `promotionalOffer.displayOnDashboard`.

File: packages/manager/src/features/ObjectStorage/ObjectStorageLanding.tsx [hunk 2]
```diff
@@ -136,6 +148,7 @@ export const ObjectStorageLanding: React.FunctionComponent<CombinedProps> = prop
           ))}
         </Tabs>
       </AppBar>
+      <PromotionalOfferCard {...promotionalOffer} fullWidth />
```
Comment (johnwcallahan) [inline packages/manager/src/features/ObjectStorage/ObjectStorageLanding.tsx]: TODO: This needs to take in to account `promotionalOffer.feature`.

File: packages/manager/src/components/PromotionalOfferCard/PromotionalOfferCard.tsx [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+export default PromotionalOfferCard;
+
+// Be extra-cautious when accessing fields on promotionalOffers, since they are
+// sourced from our external feature flag service. This function ensures that
+// each field is the type consumers are expecting, subbing defaults if they
+// aren't defined.
+export const promotionalOfferOrDefaults = (
+  offer: PromotionalOffer
+): PromotionalOffer => ({
+  name: offer.name ?? '',
```
[the remark this answers is not in the document]
Comment (johnwcallahan) [reply packages/manager/src/components/PromotionalOfferCard/PromotionalOfferCard.tsx]: Good idea, done.

**Comment by johnwcallahan** [comment]
> How confident are we that future promotions will fit this template?

@Jskobos the idea is that _eventually_ we'll have a CMS/API for these promotions. This PR describes the data shape we need to display a promotion (and the logic to do so), so that we can eventually source the content from the CMS/API instead of LD.

With this first promo, we _could_ just hardcode the content and hide behind a feature flag until the promo is active, but I think if we're going to go ahead and render based on a feature flag, we might as well start building up the structures required to source from a proper CMS.

Good catch on the colors. I updated them so they are consistent across light and dark theme.

File: packages/manager/src/components/PromotionalOfferCard/PromotionalOfferCard.tsx [hunk 4]
```diff
@@ -83,21 +97,40 @@ const PromotionalOfferCard: React.FC<CombinedProps> = props => {
...
+        {!props.fullWidth && offer.buttons && (
+          <div className={classes.buttonSection}>
+            {/* Only display the first two buttons. Any offer containing more
+            than two buttons is a mistake. */}
+            {offer.buttons.slice(0, 2).map(button => (
+              <Button
+                key={button.text}
+                className={classes.button}
+                // Check URL validity first as a security measure.
+                href={isURLValid(button.href) ? button.href : undefined}
```
Comment (Jskobos) [inline packages/manager/src/components/PromotionalOfferCard/PromotionalOfferCard.tsx]: The "cost estimator" is an external link that doesn't open in a new tab. I prefer it this way! and it's probably a limitation of the API we're using. Just pointing out that it's not consistent with what we do elsewhere

Comment (johnwcallahan) [reply packages/manager/src/components/PromotionalOfferCard/PromotionalOfferCard.tsx]: I thought about this one for a while, thought same-tab links would be simpler. Glad you like it.

**Pull request merged**